### ISSUE TITLE
[css-cascade][css-scoping] Allow layers to use different names in different contexts

### ISSUE THREAD

**Issue by knowler** [w3c/csswg-drafts#10091]
## Background

There’s some desire to write “isomorphic stylesheets” that would work for both a document and in shadow roots (https://github.com/WICG/webcomponents/issues/909, https://github.com/WICG/webcomponents/issues/986). This can be problematic, especially with layers since they are often context-specific conventions. Requiring a consumer of a web component to use the same layer names is problematic (it makes the component hard to use), just as expecting the author of a web component to know what layers a consumer is using (enforcing a layering convention on consumers seems unreasonable).

This proposal isn’t focused on how external styles can _access_ a shadow root, but rather seeks to improve the adaptability of layering when styles are used in different contexts.

## Syntax

The proposed syntax would extend the existing layering syntax:

```css
/* As a layer statement rule (useful when sharing entire stylesheets) */
@layer
  defaults as (reset in one-element, base in another-element),
  components as (elements in one-element, blocks in another-element);

/* As a layer block rule (useful when applying a subset of styles to another context) */
@layer defaults as (reset in one-element, base in another-element) {

}
```

The syntax structure would be:

```css
@layer <layer-name> as (<inner-context-layer-name> in <selector-with-inner-context>)
```

The selector would effectively match with `:host()`, e.g. if `:host(another-element)` use `base` as the layer name.

If no layer name for a specific inner context is given, the original layer name would be used.

If a context has already established its layering order, when subsequent layer statement rules appear with contextual layer names, those layer names become associated with the corresponding names in the existing order (they do not modify the order).

## Use cases

Ideally, this proposal would be complementary to an happy-path way to apply external styles to a shadow root either as a whole or in part. Even without that happy-path, I do believe it can be useful for web component authors who already are supporting ways to inject external styles into their shadow roots.

**Comment by mayank99** [comment]
It might be more useful to allow a generic way to create an alias of a layer name. It doesn't necessarily need to be tied to contexts.

```css
@layer reset as base;
```

This would be useful even without shadow DOM, especially when working with third-party stylesheets that have their own layer names. For example, `tailwindcss` v4 preconfigures `base` and `utilities` as layer names and you might want to merge those layers with your own layers (which probably have different names).

This same concept can be applied to shadow contexts by nesting the layer alias statement inside a selector (making the `in` syntax redundant).

```css
:host(one-element) {
  @layer defaults as reset;
}
:host(another-element) {
  @layer defaults as base;
}
```

I'm not sure if this fits into your proposal or should be its own separate thing, but I just wanted to mention it while it's still fresh in my mind.

**Comment by mirisuzanne** [comment]
Yes, stepping back from the specifics of shadow-dom, this seems like an identical issue to handling layer names from third-party tools. The tool might do layering internally, and those layer names may not match the names used in my project. 

The current solution is that we can wrap the tool layers into any layer of our own, when importing:

```css
@import 'tool.css' layer(my-layer-name);
```

That works great as long as we're dealing with an entire import as a single layer. Where it gets complicated is if we want to rename individual layers from the import. The `@sheet` rule could help us with that, as long as the tool provides distinct sheets for us to import into distinct layers.

If we need a solution that's more direct (without going through sheets) it still seems to me like on-import is the right time to handle that renaming. I'm a bit skeptical about creating arbitrary aliases at any point in the document, especially inside selectors.
